frontend/js/src/playlists/components/ImportPlaylistModal.tsx

```diff
@@ -0,0 +1,204 @@
+          message="You must be logged in for this operation"
+        />,
+        { toastId: "auth-error" }
+      );
+      return undefined;
+    }
+    try {
+      if (newPlaylist) {
+        const newPlaylistId = await APIService.createPlaylist(
+          currentUser?.auth_token,
+          newPlaylist
+        );
+
+        toast.success(
+          <ToastMsg
+            title="Created playlist"
+            message={
+              <>
+                Created a new playlist with ID:
+                <a href={`/playlist/${newPlaylistId}`}>{newPlaylistId}</a>
+              </>
+            }
+          />,
+          { toastId: "create-playlist-success" }
+        );
+        try {
+          // Fetch the newly created playlist and return it
+          const response = await APIService.getPlaylist(
+            newPlaylistId,
+            currentUser.auth_token
+          );
+          const JSPFObject: JSPFObject = await response.json();
+          return JSPFObject.playlist;
+        } catch (error) {
+          console.error(error);
+          return newPlaylist.playlist;
+        }
+      } else {
+        return undefined;
+      }
+    } catch (error) {
+      toast.error(
+        <ToastMsg
+          title="Could not create playlist"
+          message={`Something went wrong: ${error.toString()}`}
+        />,
+        { toastId: "create-playlist-error" }
+      );
+      return undefined;
+    }
+  }, [currentUser, newPlaylist, APIService]);
+
+  const checkFileContent = (selectedFile: File | undefined): void => {
+    // Checks if the file is a valid JSON file
+    // Note: This function does not check if the file is a valid JSPF file
+    const reader = new FileReader();
+
+    reader.onload = (e: ProgressEvent<FileReader>): void => {
+      try {
+        // If valid JSON, set the new playlist state
+        const jsonData = JSON.parse(e.target?.result as string);
+        setNewPlaylist(jsonData);
+      } catch (error) {
+        setFileError("Error parsing JSON: Please select a valid JSON file.");
+      }
+    };
+    if (selectedFile) {
+      reader.readAsText(selectedFile);
+    }
+  };
+
+  const handleFileChange = (
+    event: React.ChangeEvent<HTMLInputElement>
+  ): void => {
+    const selectedFile = event.target.files?.[0];
+
+    // Clear previous error when a new file is selected
+    setFileError(null);
+
+    // Check if the file has a valid extension
+    if (selectedFile) {
+      const fileExtension = selectedFile?.name.split(".").pop()?.toLowerCase();
+      if (fileExtension !== "jspf" && fileExtension !== "json") {
+        setFileError("Invalid file format. Please select a valid file.");
+
+        return;
+      }
+      // Check if the file content is in valid JSON format and set the new playlist state
+      checkFileContent(selectedFile);
+    }
+  };
+
+  const onSubmit = async (event: React.SyntheticEvent) => {
+    try {
+      createPlaylist();
+      modal.resolve(newPlaylist);
```

Comment:
Here you're returning the uploaded JSON directly rather than the actual playlist that we fetch from the server on lines 61-65.
I suppose `createPlaylist` should really return the playlist fetched from the server, and you can use that here instead of the uploaded file.
```
const newPlaylist = await createPlaylist();
```
^currently `newPlaylist` is taken but I commented above about renaming it, so it should be free after that